### PR TITLE
Example for Custom Metrics calculation during Federated Learning

### PR DESCRIPTION
**Pull request by gubertoli**
## Issue
A common question on the flower's questions channel from Slack is related with retrieving other metrics in addition to those available on the ML frameworks. 

### Description
This PR aims to provide a clear example on how to retrieve custom and additional metrics, to serve as baseline for other people using the Flower for Federated Learning.

### Related issues/PRs
N/A

## Proposal
A new example on how to calculate custom metrics during Federated Learning.

### Explanation
This simple example demonstrate how to calculate custom metrics over multiple clients beyond the traditional ones available in the ML frameworks. In this case, it demonstrate the use of ready-available scikit-learn metrics: accuracy, recall, precision, and f1-score.

Once both the test values (`y_test`) and the predictions (`y_pred`) are available on the client side (`client.py`), other metrics or custom ones are possible to be calculated.

The main takeaways of this implementation are:
- the use of the `output_dict` on the client side - inside `evaluate` method on `client.py`
- the use of the `evaluate_metrics_aggregation_fn` - to aggregate the metrics on the server side, part of the `strategy` on `server.py`

This example is based on the `quickstart_tensorflow` with CIFAR-10, source [here](https://flower.dev/docs/quickstart-tensorflow.html).

### Checklist

- [X] Implement proposed change
- [ ] Write tests
- [ ] Update [documentation](https://flower.dev/docs/writing-documentation.html)
- [ ] Update [changelog](https://github.com/adap/flower/blob/main/doc/source/changelog.rst)
- [ ] Make CI checks pass
- [X] Ping maintainers on [Slack](https://flower.dev/join-slack/) (channel `#contributions`)

### Any other comments?
N/A

### CODE REVIEW

**Comment by gubertoli** [comment]
@danieljanes 
Just a heads up regarding this example, as I missed the history in the Slack channel (+90 days).

File: examples/custom-metrics/requirements.txt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+flwr==1.4.0
```
Comment (danieljanes) [inline examples/custom-metrics/requirements.txt:1]:
```suggestion
flwr>=1.0, <2.0
```

File: examples/custom-metrics/pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+requires = ["poetry-core>=1.4.0"]
+build-backend = "poetry.core.masonry.api"
+
+[tool.poetry]
+name = "custom-metrics"
+version = "0.1.0"
+description = "Federated Learning with Flower and Custom Metrics"
+authors = [
+	"The Flower Authors <hello@flower.dev>",
+	"Gustavo Bertoli <>"
```
Comment (danieljanes) [inline examples/custom-metrics/pyproject.toml:11]: Add email?

File: examples/custom-metrics/client.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    rec = recall_score(
+        y_test, y_pred, average="micro"
+    )  # average argument required for multi-class
+    prec = precision_score(y_test, y_pred, average="micro")
+    f1 = f1_score(y_test, y_pred, average="micro")
+    return acc, rec, prec, f1
+
+
+# Define Flower client
+class CifarClient(fl.client.NumPyClient):
```
Comment (danieljanes) [inline examples/custom-metrics/client.py:31]: We started to default to calling these `FlowerClient`:
```suggestion
class FlowerClient(fl.client.NumPyClient):
```

File: examples/custom-metrics/client.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+
+# Make TensorFlow log less verbose
+os.environ["TF_CPP_MIN_LOG_LEVEL"] = "3"
+
+
+# Load model and data (MobileNetV2, CIFAR-10)
+model = tf.keras.applications.MobileNetV2((32, 32, 3), classes=10, weights=None)
+model.compile("adam", "sparse_categorical_crossentropy", metrics=["accuracy"])
+(x_train, y_train), (x_test, y_test) = tf.keras.datasets.cifar10.load_data()
```
Comment (danieljanes) [inline examples/custom-metrics/client.py:16]: Should we use Flower Datasets here?

File: examples/custom-metrics/pyproject.toml (hunk 2 above)
Comment (gubertoli) [reply examples/custom-metrics/pyproject.toml]: Added e-mail.

File: examples/custom-metrics/client.py (hunk 4 above)
Comment (gubertoli) [reply examples/custom-metrics/client.py]: https://github.com/adap/flower/pull/1958/commits/41043b74fe6921f4633cdc5722802ee336c9a5b8#diff-ecd736289c63680fe85bba7c7a5821ecf2a4fbd519f50c9f395c343372d9e87dR17-R27

File: examples/custom-metrics/pyproject.toml (hunk 2 above)
Comment (gubertoli) [reply examples/custom-metrics/pyproject.toml]: Added the email

File: examples/custom-metrics/client.py (hunk 3 above)
Comment (gubertoli) [reply examples/custom-metrics/client.py]: Updated the code from `CifarClient` to `FlowerClient`.

**Review by yan-gao-GY — CHANGES_REQUESTED** [review]

**Comment by yan-gao-GY** [comment]
Hi @gubertoli, thanks for your contribution for this example! I left a few suggested changes and comments.

File: examples/custom-metrics/server.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
+import flwr as fl
+import numpy as np
+
+
+def average_metrics(metrics):
```
[the remark this answers is not in the document]
Comment (gubertoli) [reply examples/custom-metrics/server.py:6]: I addressed this comment by creating a more descriptive docstring [here](https://github.com/adap/flower/pull/1958/commits/e5ecd6b483d3151bf1e3e50739fde2f072ee57a3#diff-a4cac5fbf18766f5ef2af0bfbe981857c1642507f6bce73ecb82a8d9374b59b1R7-R33):

```python
# Define metrics aggregation function
def average_metrics(metrics):
    """Aggregate metrics from multiple clients by calculating mean averages.

    Parameters:
    - metrics (list): A list containing tuples, where each tuple represents metrics for a client.
                    Each tuple is structured as (num_examples, metric), where:
                    - num_examples (int): The number of examples used to compute the metrics.
                    - metric (dict): A dictionary containing custom metrics provided as `output_dict`
                                    in the `evaluate` method from `client.py`.

    Returns:
    A dictionary with the aggregated metrics, calculating mean averages. The keys of the
    dictionary represent different metrics, including:
    - 'accuracy': Mean accuracy calculated by TensorFlow.
    - 'acc': Mean accuracy from scikit-learn.
    - 'rec': Mean recall from scikit-learn.
    - 'prec': Mean precision from scikit-learn.
    - 'f1': Mean F1 score from scikit-learn.

    Note: If a weighted average is required, the `num_examples` parameter can be leveraged.

    Example:
        Example `metrics` list for two clients after the last round:
        [(10000, {'prec': 0.108, 'acc': 0.108, 'f1': 0.108, 'accuracy': 0.1080000028014183, 'rec': 0.108}),
        (10000, {'f1': 0.108, 'rec': 0.108, 'accuracy': 0.1080000028014183, 'prec': 0.108, 'acc': 0.108})]
    """

    # Here num_examples are not taken into account by using _
    accuracies_tf = np.mean([metric["accuracy"] for _, metric in metrics])
    accuracies = np.mean([metric["acc"] for _, metric in metrics])
    recalls = np.mean([metric["rec"] for _, metric in metrics])
    precisions = np.mean([metric["prec"] for _, metric in metrics])
    f1s = np.mean([metric["f1"] for _, metric in metrics])

    return {
        "accuracy": accuracies_tf,
        "acc": accuracies,
        "rec": recalls,
        "prec": precisions,
        "f1": f1s,
    }
```

File: examples/custom-metrics/client.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+os.environ["TF_CPP_MIN_LOG_LEVEL"] = "3"
+
+
+# Load model (MobileNetV2)
+model = tf.keras.applications.MobileNetV2((32, 32, 3), classes=10, weights=None)
+model.compile("adam", "sparse_categorical_crossentropy", metrics=["accuracy"])
+
+# Load data with Flower Datasets (CIFAR-10)
+fds = FederatedDataset(dataset="cifar10", partitioners={"train": 10})
+train = fds.load_full("train")
```
[the remark this answers is not in the document]
Comment (gubertoli) [reply examples/custom-metrics/client.py:20]: If you have any suggestions on this specific part, I would like to change it. However, the focus is on the additional metrics calculation.

**Review by yan-gao-GY — APPROVED** [review]

**Review by danieljanes — APPROVED** [review]

**Comment by danieljanes** [comment]
Thanks for the contribution @gubertoli!
